### PR TITLE
fixes #2734: improve iOS testing

### LINKED ISSUE #2734: iOS testing failures around location permissions
We had hangs on Travis during tests because the location permission dialog was going untapped (since #2612), which we addressed in #2645. This was tweaked somewhat in #2685 to speed things up (as waiting for the permission dialog takes 10 seconds).

There have been some different failures since #2685:

```
Test Case '-[MapViewTests testDelegatesStartStopLocatingUser]' started.
2015-10-22 13:47:54.330 Mapbox GL Tests[3963:13037] _BSMachError: (os/kern) invalid capability (20)
2015-10-22 13:47:54.334 Mapbox GL Tests[3963:13093] _BSMachError: (os/kern) invalid name (15)
2015-10-22 13:47:54.385 Mapbox GL Tests[3963:13037] _BSMachError: (os/kern) invalid capability (20)
2015-10-22 13:47:57.477 Mapbox GL Tests[3963:13037] tap point is not within the bounds of the screen
/Users/vagrant/git/test/ios/KIF/Classes/UIAutomationHelper.m:61: error: -[MapViewTests testDelegatesStartStopLocatingUser] : failed: caught "UIAExceptionBadPoint", "tap point is not within the bounds of the screen"
(
	0   CoreFoundation                      0x0000000105e4bf65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001058c3deb objc_exception_throw + 48
	2   CoreFoundation                      0x0000000105e4be9d +[NSException raise:format:] + 205
	3   UIAutomation                        0x000000011c0992a3 -[UIATarget _tapOptionalObject:tapCount:touchCount:tapOffset:] + 466
	4   UIAutomation                        0x000000011c087f4b -[UIAElement tap] + 319
	5   Test Bundle                         0x00000001155d2835 -[UIAutomationHelper acknowledgeSystemAlert] + 203
	6   Test Bundle                         0x00000001155d26e4 +[UIAutomationHelper acknowledgeSystemAlert] + 48
	7   Test Bundle                         0x00000001155b739c -[MapViewTests approveLocationIfNeeded] + 72
	8   Test Bundle                         0x00000001155de186 -[KIFSystemTestActor waitForNotificationName:object:whileExecutingBlock:complete:] + 360
	9   Test Bundle                         0x00000001155ddfea -[KIFSystemTestActor waitForNotificationName:object:whileExecutingBlock:] + 74
	10  Test Bundle                         0x00000001155cc77b -[MapViewTests testDelegatesStartStopLocatingUser] + 313
	11  CoreFoundation                      0x0000000105d3a85c __invoking___ + 140
	12  CoreFoundation                      0x0000000105d3a6ae -[NSInvocation invoke] + 286
	13  XCTest                              0x000000010fa2c608 __24-[XCTestCase invokeTest]_block_invoke_2 + 159
	14  XCTest                              0x000000010fa5bc90 -[XCTestContext performInScope:] + 184
	15  XCTest                              0x000000010fa2c558 -[XCTestCase invokeTest] + 169
	16  XCTest                              0x000000010fa2c9f3 -[XCTestCase performTest:] + 443
	17  XCTest                              0x000000010fa2a6b9 -[XCTestSuite performTest:] + 377
	18  XCTest                              0x000000010fa2a6b9 -[XCTestSuite performTest:] + 377
	19  XCTest                              0x000000010fa2a6b9 -[XCTestSuite performTest:] + 377
	20  XCTest                              0x000000010fa17ec2 __25-[XCTestDriver _runSuite]_block_invoke + 51
	21  XCTest                              0x000000010fa380cb -[XCTestObservationCenter _observeTestExecutionForBlock:] + 615
	22  XCTest                              0x000000010fa17e0e -[XCTestDriver _runSuite] + 408
	23  XCTest                              0x000000010fa18b6d -[XCTestDriver _checkForTestManager] + 263
	24  XCTest                              0x000000010fa5cff1 _XCTestMain + 628
	25  CoreFoundation                      0x0000000105d7808c __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
	26  CoreFoundation                      0x0000000105d6dd15 __CFRunLoopDoBlocks + 341
	27  CoreFoundation                      0x0000000105d6d472 __CFRunLoopRun + 850
	28  CoreFoundation                      0x0000000105d6ce98 CFRunLoopRunSpecific + 488
	29  GraphicsServices                    0x000000010bd73ad2 GSEventRunModal + 161
	30  UIKit                               0x00000001067b7676 UIApplicationMain + 171
	31  Mapbox GL Tests                     0x000000010483a0d3 main + 80
	32  libdyld.dylib                       0x000000010883a92d start + 1
)
Test Case '-[MapViewTests testDelegatesStartStopLocatingUser]' failed (7.664 seconds).
```

[and](https://travis-ci.org/mapbox/mapbox-gl-native/builds/86842717)

```
MapViewTests
  testUserTrackingModeFollow, ((tester.mapView.userLocationVisible) equal to (YES)) failed: ("NO") is not equal to ("YES") - user location should be visible
  /Users/travis/build/mapbox/mapbox-gl-native/test/ios/MapViewTests.m:630
                   YES,
                   @"user location should be visible");
    XCTAssertEqual(tester.mapView.userLocation.coordinate.latitude,
 ```

[and](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/86788911)

```
MapViewTests
  testDelegatesStartStopLocatingUser, The step timed out after 10.00 seconds: Application is ignoring interaction events
  /Users/travis/build/mapbox/mapbox-gl-native/test/ios/KIFTestActor+MapboxGL.m:18

- (MGLMapView *)mapView {
    return (MGLMapView *)[tester waitForViewWithAccessibilityLabel:@"Map"];
}

  testDelegatesStartStopLocatingUser, Failure in child step: The step timed out after 10.00 seconds: Application is ignoring interaction events
  /Users/travis/build/mapbox/mapbox-gl-native/test/ios/KIFTestActor+MapboxGL.m:14

- (MGLTViewController *)viewController {
    return (MGLTViewController *)[[tester.mapView nextResponder] nextResponder];
}

  testDelegatesStartStopLocatingUser, Failure in child step: The step timed out after 10.00 seconds: Application is ignoring interaction events
  /Users/travis/build/mapbox/mapbox-gl-native/test/ios/MapViewTests.m:27

    [system simulateDeviceRotationToOrientation:UIDeviceOrientationPortrait];
    [tester.viewController resetMapView];

	 Executed 16 tests, with 3 failures (3 unexpected) in 32.161 (32.182) seconds
```

/cc @incanus @1ec5

### PR DESCRIPTION
**Pull request by incanus**
- Got rid of the heading tracking test. Mocking this is proving too complicated (plus I turned up two GL library bugs in the process) and we do builds often enough to keep manual tabs on this functionality. I just don't think it's worth the hassle to automate, especially given the low risk of functionality breakage. 

- Fixed said tracking-related bugs in `MGLMapView`. 

- Removed the external location mocking library to make this more like other test suites. 

- Moved to `Debug` since it runs faster these days. 

- Test tweaks for more reliability. 

- Enabled all tests. In future, if tests are flapping, we should remove them and ticket replacement, not disable them. 

Going to squash/rebase/finalize etc. before merging. 

/cc @friedbunny @1ec5

### CODE REVIEW

File: platform/ios/MGLMapView.mm [hunk 1]
```diff
@@ -2648,7 +2641,11 @@ - (void)setUserTrackingMode:(MGLUserTrackingMode)mode animated:(BOOL)animated
         {
             self.showsUserLocation = YES;
 
-            if (self.zoomLevel < self.currentMinimumZoom) [self setZoomLevel:self.currentMinimumZoom animated:YES];
+            if (self.zoomLevel < self.currentMinimumZoom)
+            {
+                [self setZoomLevel:self.currentMinimumZoom animated:YES];
+                _userTrackingMode = MGLUserTrackingModeFollowWithHeading; // reapply
```
Comment (1ec5) [inline platform/ios/MGLMapView.mm]: Good idea.

**Comment by friedbunny** [comment]
Happy to be rid of the old mock location stuff, was clunky. I'll kick the tires on this a bit now...

**Comment by friedbunny** [comment]
Travis is failing on `testCompassTap`:

```
MapViewTests
  testCompassTap, ((tester.mapView.direction) equal to (180)) failed: ("333.527575343") is not equal to ("180") - setting direction should take effect
  /Users/travis/build/mapbox/mapbox-gl-native/test/ios/MapViewTests.m:99

                   180,
                   @"setting direction should take effect");

	 Executed 22 tests, with 1 failure (0 unexpected) in 39.952 (39.976) seconds
```

And bitrise is failing on `testUserTrackingModeFollow`:

```
/Users/vagrant/git/test/ios/MapViewTests.m:651: error: -[MapViewTests testUserTrackingModeFollow] : ((tester.mapView.userLocation.coordinate.latitude) equal to (kMockedLatitude)) failed: ("37.785834") is not equal to ("37.775716") - user location latitude should match mocked latitude
Test Case '-[MapViewTests testUserTrackingModeFollow]' failed (3.688 seconds).
```

**Comment by friedbunny** [comment]
Seeing `testUserTrackingModeFollow` fail locally, as well:

```
testUserTrackingModeFollow, ((tester.mapView.userLocation.coordinate.latitude) equal to (kMockedLatitude)) failed: ("37.33233141") is not equal to ("37.775716") - user location latitude should match mocked latitude
/mapbox-gl-native/test/ios/MapViewTests.m:651
```

**Comment by incanus** [comment]
Yeah, still getting fails here, hardly ever locally, but flapping on Bitrise and Travis. Just rebased off of a243f41cb4807aa3784e2133986bda270464517c to ditch Travis, though.

**Comment by incanus** [comment]
Current take here is to: 

- Remove user tracking-related tests as just too unstable; ticket as todo. 
- Remove currently-disabled metrics tests; ticket as todo. 
- Rebase, squash, etc. and merge this sucker. 

Refs https://github.com/mapbox/mapbox-gl-native/pull/2213

**Comment by incanus** [comment]
This is passing Bitrise and the iOS slice is no longer on Travis, so will prep this for merge now.

**Comment by 1ec5** [comment]
Still getting this failure locally:

```
MapViewTests
  testCompassTap, ((tester.mapView.direction) equal to (180)) failed: ("226.279000419") is not equal to ("180") - setting direction should take effect
  /path/to/mapbox-gl-native/test/ios/MapViewTests.m:70
                   180,
                   @"setting direction should take effect");

```

**Comment by 1ec5** [comment]
The failure is being tracked in #2920.

**Pull request merged**